src/couch_httpd_csrf.erl

```diff
@@ -0,0 +1,172 @@
+% use this file except in compliance with the License.  You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+%% This module provides optional CSRF protection to any client
+%%
+%% Clients should use the following pseudo code;
+%% if (hasCookie("CouchDB-CSRF")) {
+%%   setRequestHeader("X-CouchDB-CSRF", cookieValue("CouchDB-CSRF"));
+%% } else {
+%%   setRequestHeader("X-CouchDB-CSRF", "true")
+%% }
+%%
+%% If CouchDB sees the CouchDB-CSRF cookie then it checks its validity
+%% and whether the X-CouchDB-CSRF request header exists and matches.
+%% A 403 is returned if those checks fail.
+%% If CouchDB does not see the CouchDB-CSRF cookie but does see
+%% the X-CouchDB-CSRF header with value "true", a CouchDB-CSRF cookie
+%% is generated and returned.
+
+-module(couch_httpd_csrf).
+
+-export([validate/1, headers/2]).
+
+-include_lib("couch/include/couch_db.hrl").
+
+validate(#httpd{} = Req) ->
+    Cookie = csrf_from_req(Req),
+    Header = couch_httpd:header_value(Req, "X-CouchDB-CSRF"),
+    case {Cookie, Header} of
+        {undefined, undefined} ->
+            ok;
+        {undefined, "true"} ->
+            ok;
+        {undefined, _} ->
+            throw({forbidden, <<"CSRF header sent without cookie">>});
+        {Csrf, Csrf} ->
+            %% cookie and header match, but is it valid?
+            ok = validate(Csrf),
+
+            %% check expiration
+            {_, Timestamp, _} = decode_cookie(Csrf),
+            MaxAge = max_age(),
+            case Timestamp + MaxAge < timestamp() of
+                true ->
+                    throw({forbidden, <<"CSRF Cookie expired">>});
+                false ->
+                    ok
+            end;
+        _ ->
+            throw({forbidden, <<"CSRF Cookie/Header mismatch">>})
+    end;
+%% Check that we generated this CSRF token
+validate(Csrf) when is_list(Csrf) ->
+    case decode_cookie(Csrf) of
+        invalid ->
+            throw({bad_request, <<"Malformed CSRF Cookie">>});
+        Cookie ->
+            case validate_cookie(Cookie) of
+                true ->
+                    ok;
+                false ->
+                    throw({forbidden, <<"Invalid CSRF value">>})
+            end
+    end.
+
+
+headers(#httpd{} = Req, Headers) ->
+    Header = couch_httpd:header_value(Req, "X-CouchDB-CSRF"),
+    case {csrf_from_req(Req), csrf_in_headers(Headers), Header} of
+        {undefined, false, "true"} ->
+            [make_cookie() | Headers];
+        {Csrf, false, Csrf} when Csrf /= undefined ->
+            case decode_cookie(Csrf) of
+                invalid ->
+                    [make_cookie() | Headers];
```

Comment:
Is there a case for this branch? `validate/1` calls in `handle_request_int` which will throw `bad_request` if `Csrf` is invalid while I see `headers/1` is only called in response when `Csrf` may be `undefined` or valid cookie as otherwise it will not pass validation. Or I miss something?